SwiftCharts/Views/ChartAreasView.swift

```diff
@@ -11,92 +11,68 @@ import UIKit
-            progressline.addLine(to: p)
-        }
-        
-        progressline.close()
-        
-        return progressline
+    
+    fileprivate func generateAreaPath(points: [CGPoint], pathGenerator: ChartLinesViewPathGenerator) -> UIBezierPath {
+        return pathGenerator.generateAreaPath(points: points, lineWidth: 1.0)
     }
     
     fileprivate func show(path: UIBezierPath) {
         guard let firstColor = colors.first else {
             print("WARNING: No color(s) used for ChartAreasView")
             return
         }
-        let areaLayer = CAShapeLayer()
-        areaLayer.lineJoin = kCALineJoinBevel
-        areaLayer.fillColor   = firstColor.cgColor
-        areaLayer.lineWidth   = 2.0
-        areaLayer.strokeEnd   = 0.0
+        
         if colors.count == 1 {
-            layer.addSublayer(areaLayer)
+            colors.append(colors[0])
         }
         
-        areaLayer.path = path.cgPath
-        areaLayer.strokeColor = firstColor.cgColor
+        let shape = CAShapeLayer()
+        shape.frame = self.bounds
         
-        if colors.count > 1 {
-            let gradient = CAGradientLayer()
-            gradient.anchorPoint = CGPoint.zero
-            var CGColors: [CGColor] = []
-            for color in colors {
-                CGColors.append(color.cgColor)
-            }
-            gradient.colors = CGColors
-            gradient.bounds = layer.bounds
-            
-            let pathY = path.bounds.origin.y
-            let pathHeight = path.bounds.height
-            let maskHeight = gradient.bounds.height
+        shape.path = path.cgPath
+        shape.strokeColor = firstColor.cgColor
+        shape.lineWidth = 2
+        shape.fillColor = nil
         
-            var bottomY: CGFloat?
-            var topY: CGFloat?
-            if pathY < pathHeight {
-                bottomY = (pathY/pathHeight)/2
-                topY = ((maskHeight - (pathHeight-pathY))/maskHeight)/2
-            } else {
-                bottomY = 0
-                topY = ((maskHeight-pathHeight)/maskHeight)
-            }
-            
-            gradient.startPoint = CGPoint(x: 0.5, y: 0+topY!)
-            gradient.endPoint = CGPoint(x: 0.5, y: 1.0-bottomY!)
-            gradient.mask = areaLayer
-            layer.addSublayer(gradient)
+        let gradient = CAGradientLayer()
+        gradient.frame = self.bounds
+        gradient.colors = colors.map{$0.cgColor}
+        
+        let mask = CAShapeLayer()
+        mask.frame = self.bounds
+        
+        if addContainerPoints {
+            path.addLine(to: CGPoint(x: shape.frame.size.width, y: shape.frame.size.height))
+            path.addLine(to: CGPoint(x: 0, y: shape.frame.size.height))
+            path.close()
         }
         
+        mask.path = path.cgPath
+        mask.fillColor = UIColor.black.cgColor
+        gradient.mask = mask
+        
+        self.layer.addSublayer(gradient)
+        self.layer.addSublayer(shape)
```

Comment:
is `self` necessary in these statements?

Reply:
removed self from these two calls